src/select-query-parser/result.ts

```diff
@@ -125,12 +125,12 @@ export type ProcessNode<
   RelationName extends string,
   Relationships extends GenericRelationship[],
   NodeType extends Ast.Node
-> = NodeType extends Ast.StarNode // If the selection is *
+> = NodeType['type'] extends Ast.StarNode['type'] // If the selection is *
   ? Row
-  : NodeType extends Ast.SpreadNode // If the selection is a ...spread
-  ? ProcessSpreadNode<Schema, Row, RelationName, Relationships, NodeType>
-  : NodeType extends Ast.FieldNode
-  ? ProcessFieldNode<Schema, Row, RelationName, Relationships, NodeType>
+  : NodeType['type'] extends Ast.SpreadNode['type'] // If the selection is a ...spread
+  ? ProcessSpreadNode<Schema, Row, RelationName, Relationships, Extract<NodeType, Ast.SpreadNode>>
+  : NodeType['type'] extends Ast.FieldNode['type']
+  ? ProcessFieldNode<Schema, Row, RelationName, Relationships, Extract<NodeType, Ast.FieldNode>>
```

Comment:
**remark**
User signaled from discord an issue with types inference in some cases when using spread operator: https://discord.com/channels/839993398554656828/1300910943634460682

This was due to the fact that `NodeType extends Ast.SpreadNode` will also returns true if the `NodeType` is actually a `Ast.FieldNode` even if their `type` discriminator are different.

I couldn't really understand exactly why forcing the check over the type property itself work while checking the whole type didn't but it fixes this issue.

Note that I couldn't reproduce the issue with our own test database, but the test based on the user report I used for this can be found in this commit: https://github.com/supabase/postgrest-js/pull/569/commits/34c9c827f85ad10a1dd09419799c96def881b1cf

Reply:
Not sure if these changes are needed - tests pass without them, only the change in `ProcessSpreadNodeResult` is needed.

Reply:
Yeah this one is a nasty one, and as I mentioned I'm not completely sure of what's happening but I finally managed to get this reproduced in a test case here:

https://github.com/supabase/postgrest-js/pull/569/commits/071185a40f929289fdaab44782e5382bc60b671e#diff-db137d8db955b42a3a2e437ed10e24dcdd2e2754c5116a3b0259aebb681f167aR104-R120

Here is the debugging steps that I took to get to the fix. Basically the root cause of the error, is that, at some point, `ProcessNode` fail to infer the right node type I still don't understand why. But at some point a `FieldNode` is sent to the `SpreadNode` processing function. And here we can see that this is because this "Node extends Ast.SpreadNode" returns `true`. 

<img width="1052" alt="Screenshot 2024-10-30 at 11 09 33" src="https://github.com/user-attachments/assets/b6e205ac-abca-43e6-bd3e-fe82d7f10b0a">

Note that, inverting the conditional order in which we check the conditions (checking the `FieldNode` before `SpreadNode`) in `ProcessNode` would also fix this case, because the node actually result in `true` for BOTH types.

<img width="1115" alt="Screenshot 2024-10-30 at 11 28 29" src="https://github.com/user-attachments/assets/5a6a2541-aec0-4297-bf51-4512dd0510d7">

Where it becomes really really weird, is that reproducing the exact same node in a literal and testing the assertion does not give the same result:

<img width="438" alt="Screenshot 2024-10-30 at 11 11 00" src="https://github.com/user-attachments/assets/a9f4d5cf-112e-48f3-85dd-d66e78cd3088">

Here the `extends Ast.SpreadNode` properly result in `false` as expected.

My solution is to make sure in `ProcessNode` that we discriminate over the `type` property of the node. Doing so the SpreadNode processing is no longer called over a `FieldNode` object:


<img width="1094" alt="Screenshot 2024-10-30 at 11 11 54" src="https://github.com/user-attachments/assets/5bc32cfb-d832-4bb9-90b8-7eea960e8788">

But as a mentioned, on this one I'm not entirely sure about WHY the `extends` check doesn't behave as intended in this specific case. So I'm open to suggestions.